export/mlab_export.py

```diff
@@ -0,0 +1,591 @@
+    if self._first_creation:
+      # Since the file does not exist, use mode 'w' to create it.
+      self._lock_handle = open(self._filename, 'w')
+    else:
+      # File *does* exist, so use mode 'r' to preserve the mtime on the file.
+      self._lock_handle = open(self._filename, 'r')
+
+  def close(self):
+    """Closes the timestamp file."""
+    if self._lock_handle is not None:
+      self._lock_handle.close()
+
+  def get_mtime(self):
+    """Returns the file mtime, or zero if file was just created."""
+    if self._first_creation:
+      # Indicate that there is no timestamp, so the caller can use a default.
+      return 0
+    return int(os.stat(self._filename).st_mtime)
+
+  def update_mtime(self, mtime):
+    """Updates the atime & mtime on the export timestamp file."""
+    os.utime(self._filename, (mtime, mtime))
+
+  def lock(self):
+    """Attempt to acquire advisory lock on export timestamp file.
+
+    This method returns immediatly.
+
+    Returns:
+      bool, True if the lock is acquired, False otherwise.
+    """
+    # Attempt an exclusive lock without blocking. On success, fcntl.flock()
+    # returns no value. On error, it throws IOError EAGAIN "temporarily
+    # unavailable."
+    try:
+      fcntl.flock(self._lock_handle, fcntl.LOCK_EX|fcntl.LOCK_NB)
+      return True
+    except IOError:
+      return False
+
+
+def align_timestamp(timestamp, step):
+  """Adjusts 'timestamp' down to a multiple of 'step' size.
+
+  Args:
+    timestamp: int, timestamp in seconds since the epoch.
+    step: int, interval to align.
+  Returns:
+    int, timestamp adjusted to multiple of step.
+  """
+  return timestamp - (timestamp % step)
+
+
+def default_end_time(options):
+  """Calculates a default, end timestamp.
+
+  Args:
+    options: flags.FlagValues, the runtime options. These values are read:
+        options.length, options.step, options.update, options.ts_offset.
+  Returns:
+    int, timestamp in seconds since the epoch.
+  """
+  ts_current = int(time.time())
+
+  # Align current ts to a multiple of length.
+  ts_end = align_timestamp(ts_current, options.length)
+
+  # Align end ts to a multiple of step size (just in case).
+  ts_end = align_timestamp(ts_end, options.step)
+
+  if options.update and (ts_current - ts_end) < options.ts_offset:
+    raise TimeOptionError('ts_current - ts_end < min offset: %s - %s < %s' %
+        (ts_current, ts_end, options.ts_offset))
+  return ts_end
+
+
+def default_start_time(options, ts_previous):
+  """Calculates a default start timestamp.
+
+  Args:
+    options: flags.FlagValues, the runtime options. These values are read:
+        options.length, options.step, options.update, options.ts_end.
+    ts_previous: int, timestamp in seconds since epoch of last
+        successful export. On first run, this value should be zero.
+  Returns:
+    int, timestamp in seconds since the epoch.
+  """
+  # Only when both options.update and ts_previous are set, skip assignment.
```

Comment:
This is confusing because the `if` itself is when _neither_ are set, the opposite of the logic of this comment. I think the comment on 254 is sufficient.